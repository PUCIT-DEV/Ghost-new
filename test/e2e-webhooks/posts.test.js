const {agentProvider, mockManager, fixtureManager, matchers} = require('../utils/e2e-framework');
const {anyObjectId, anyISODateTime, anyUuid, stringMatching} = matchers;

const tierSnapshot = {
    id: anyObjectId,
    created_at: anyISODateTime,
    updated_at: anyISODateTime
};

const buildAuthorSnapshot = (roles = false) => {
    const authorSnapshot = {
        last_seen: anyISODateTime,
        created_at: anyISODateTime,
        updated_at: anyISODateTime
    };

    // NOTE: this is such a bad hack! for the reasons I did not investigate the "add" event does not include
    //       the roles but the "published" does! massive inconsistency and needs to be fixed one day
    if (roles) {
        authorSnapshot.roles = new Array(1).fill({
            id: anyObjectId,
            created_at: anyISODateTime,
            updated_at: anyISODateTime
        });
    }

    return authorSnapshot;
};

const buildPostSnapshotWithTiers = ({published, tiersCount, roles = false}) => {
    return {
        id: anyObjectId,
        uuid: anyUuid,
        comment_id: anyObjectId,
        published_at: published ? anyISODateTime : null,
        created_at: anyISODateTime,
        updated_at: anyISODateTime,
        // @TODO: hack here! it's due to https://github.com/TryGhost/Toolbox/issues/341
        //        this matcher should be removed once the issue is solved
        url: stringMatching(/http:\/\/127.0.0.1:2369\/\w+\//),
        tiers: new Array(tiersCount).fill(tierSnapshot),
        primary_author: buildAuthorSnapshot(roles),
        authors: new Array(1).fill(buildAuthorSnapshot(roles))
    };
};

const buildPreviousPostSnapshotWithTiers = ({tiersCount}) => {
    return {
        updated_at: anyISODateTime,
        tiers: new Array(tiersCount).fill(tierSnapshot)
    };
};

describe('post.* events', function () {
    let adminAPIAgent;
    let webhookMockReceiver;

    before(async function () {
        adminAPIAgent = await agentProvider.getAdminAPIAgent();
        await fixtureManager.init('integrations');
        await adminAPIAgent.loginAsOwner();
    });

    beforeEach(function () {
        webhookMockReceiver = mockManager.mockWebhookRequests();
    });

    afterEach(function () {
        mockManager.restore();
    });

    it('post.published event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/post-published/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'post.published',
            url: webhookURL
        });

        const res = await adminAPIAgent
            .post('posts/')
            .body({
                posts: [{
                    title: 'webhookz',
                    status: 'draft'
                }]
            })
            .expectStatus(201);

        const id = res.body.posts[0].id;
        const updatedPost = res.body.posts[0];
        updatedPost.status = 'published';

        await adminAPIAgent
            .put('posts/' + id)
            .body({
                posts: [updatedPost]
            })
            .expectStatus(200);

        await webhookMockReceiver
            // TODO: implement header matching feature next!
            // .matchHeaderSnapshot();
            .matchBodySnapshot({
                post: {
                    current: buildPostSnapshotWithTiers({
                        published: true,
                        tiersCount: 2,
                        roles: true
                    }),
                    previous: buildPreviousPostSnapshotWithTiers({
                        tiersCount: 2
                    })
                }
            });
    });

    it('post.added event is triggered', async function () {
        const webhookURL = 'https://test-webhook-receiver.com/post-added/';
        await webhookMockReceiver.mock(webhookURL);
        await fixtureManager.insertWebhook({
            event: 'post.added',
            url: webhookURL
        });

        await adminAPIAgent
            .post('posts/')
            .body({
                posts: [{
                    title: 'testing post.added webhook',
                    status: 'draft'
                }]
            })
            .expectStatus(201);

        await webhookMockReceiver
            // TODO: implement header matching feature next!
            // .matchHeaderSnapshot();
            .matchBodySnapshot({
                post: {
                    current: buildPostSnapshotWithTiers({
                        published: false,
                        tiersCount: 2
                    })
                }
            });
    });
});
