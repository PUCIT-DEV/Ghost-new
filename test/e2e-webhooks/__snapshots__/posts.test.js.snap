// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post.* events post.added event is triggered 1: [body] 1`] = `
Object {
  "post": Object {
    "current": Object {
      "authors": Array [
        Object {
          "accessibility": null,
          "bio": "bio",
          "comment_notifications": true,
          "cover_image": null,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "email": "jbloggs@example.com",
          "facebook": null,
          "id": "1",
          "last_seen": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "location": "location",
          "meta_description": null,
          "meta_title": null,
          "name": "Joe Bloggs",
          "profile_image": "https://example.com/super_photo.jpg",
          "slug": "joe-bloggs",
          "status": "active",
          "tour": null,
          "twitter": null,
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "url": "http://127.0.0.1:2369/author/joe-bloggs/",
          "website": null,
        },
      ],
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "comment_id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "custom_excerpt": null,
      "custom_template": null,
      "email": Object {},
      "email_only": false,
      "email_segment": "all",
      "email_subject": null,
      "excerpt": null,
      "feature_image": null,
      "feature_image_alt": null,
      "feature_image_caption": null,
      "featured": false,
      "frontmatter": null,
      "html": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "mobiledoc": "{\\"version\\":\\"0.3.1\\",\\"ghostVersion\\":\\"4.0\\",\\"markups\\":[],\\"atoms\\":[],\\"cards\\":[],\\"sections\\":[[1,\\"p\\",[[0,[],0,\\"\\"]]]]}",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "primary_author": Object {
        "accessibility": null,
        "bio": "bio",
        "comment_notifications": true,
        "cover_image": null,
        "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
        "email": "jbloggs@example.com",
        "facebook": null,
        "id": "1",
        "last_seen": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
        "location": "location",
        "meta_description": null,
        "meta_title": null,
        "name": "Joe Bloggs",
        "profile_image": "https://example.com/super_photo.jpg",
        "slug": "joe-bloggs",
        "status": "active",
        "tour": null,
        "twitter": null,
        "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
        "url": "http://127.0.0.1:2369/author/joe-bloggs/",
        "website": null,
      },
      "primary_tag": null,
      "published_at": null,
      "slug": "testing-post-added-webhook",
      "status": "draft",
      "tags": Array [],
      "tiers": Array [
        Object {
          "active": true,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "monthly_price_id": null,
          "name": "Default Product",
          "slug": "default-product",
          "type": "paid",
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "visibility": "public",
          "welcome_page_url": null,
          "yearly_price_id": null,
        },
        Object {
          "active": true,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "monthly_price_id": null,
          "name": "Free",
          "slug": "free",
          "type": "free",
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "visibility": "public",
          "welcome_page_url": null,
          "yearly_price_id": null,
        },
      ],
      "title": "testing post.added webhook",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": StringMatching /http:\\\\/\\\\/127\\.0\\.0\\.1:2369\\\\/\\\\w\\+\\\\//,
      "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "visibility": "public",
    },
    "previous": Object {},
  },
}
`;

exports[`post.* events post.published event is triggered 1: [body] 1`] = `
Object {
  "post": Object {
    "current": Object {
      "authors": Array [
        Object {
          "accessibility": null,
          "bio": "bio",
          "comment_notifications": true,
          "cover_image": null,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "email": "jbloggs@example.com",
          "facebook": null,
          "id": "1",
          "last_seen": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "location": "location",
          "meta_description": null,
          "meta_title": null,
          "name": "Joe Bloggs",
          "profile_image": "https://example.com/super_photo.jpg",
          "roles": Array [
            Object {
              "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
              "description": "Blog Owner",
              "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
              "name": "Owner",
              "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
            },
          ],
          "slug": "joe-bloggs",
          "status": "active",
          "tour": null,
          "twitter": null,
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "url": "http://127.0.0.1:2369/author/joe-bloggs/",
          "website": null,
        },
      ],
      "canonical_url": null,
      "codeinjection_foot": null,
      "codeinjection_head": null,
      "comment_id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "custom_excerpt": null,
      "custom_template": null,
      "email_only": false,
      "email_segment": "all",
      "email_subject": null,
      "excerpt": null,
      "feature_image": null,
      "feature_image_alt": null,
      "feature_image_caption": null,
      "featured": false,
      "frontmatter": null,
      "html": null,
      "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
      "meta_description": null,
      "meta_title": null,
      "mobiledoc": "{\\"version\\":\\"0.3.1\\",\\"ghostVersion\\":\\"4.0\\",\\"markups\\":[],\\"atoms\\":[],\\"cards\\":[],\\"sections\\":[[1,\\"p\\",[[0,[],0,\\"\\"]]]]}",
      "og_description": null,
      "og_image": null,
      "og_title": null,
      "primary_author": Object {
        "accessibility": null,
        "bio": "bio",
        "comment_notifications": true,
        "cover_image": null,
        "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
        "email": "jbloggs@example.com",
        "facebook": null,
        "id": "1",
        "last_seen": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
        "location": "location",
        "meta_description": null,
        "meta_title": null,
        "name": "Joe Bloggs",
        "profile_image": "https://example.com/super_photo.jpg",
        "roles": Array [
          Object {
            "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
            "description": "Blog Owner",
            "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
            "name": "Owner",
            "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          },
        ],
        "slug": "joe-bloggs",
        "status": "active",
        "tour": null,
        "twitter": null,
        "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
        "url": "http://127.0.0.1:2369/author/joe-bloggs/",
        "website": null,
      },
      "primary_tag": null,
      "published_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "slug": "webhookz",
      "status": "published",
      "tags": Array [],
      "tiers": Array [
        Object {
          "active": true,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "monthly_price_id": null,
          "name": "Default Product",
          "slug": "default-product",
          "type": "paid",
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "visibility": "public",
          "welcome_page_url": null,
          "yearly_price_id": null,
        },
        Object {
          "active": true,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "monthly_price_id": null,
          "name": "Free",
          "slug": "free",
          "type": "free",
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "visibility": "public",
          "welcome_page_url": null,
          "yearly_price_id": null,
        },
      ],
      "title": "webhookz",
      "twitter_description": null,
      "twitter_image": null,
      "twitter_title": null,
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
      "url": StringMatching /http:\\\\/\\\\/127\\.0\\.0\\.1:2369\\\\/\\\\w\\+\\\\//,
      "uuid": StringMatching /\\[a-f0-9\\]\\{8\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{4\\}-\\[a-f0-9\\]\\{12\\}/,
      "visibility": "public",
    },
    "previous": Object {
      "published_at": null,
      "status": "draft",
      "tiers": Array [
        Object {
          "active": true,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "monthly_price_id": null,
          "name": "Default Product",
          "slug": "default-product",
          "type": "paid",
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "visibility": "public",
          "welcome_page_url": null,
          "yearly_price_id": null,
        },
        Object {
          "active": true,
          "created_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "description": null,
          "id": StringMatching /\\[a-f0-9\\]\\{24\\}/,
          "monthly_price_id": null,
          "name": "Free",
          "slug": "free",
          "type": "free",
          "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
          "visibility": "public",
          "welcome_page_url": null,
          "yearly_price_id": null,
        },
      ],
      "updated_at": StringMatching /\\\\d\\{4\\}-\\\\d\\{2\\}-\\\\d\\{2\\}T\\\\d\\{2\\}:\\\\d\\{2\\}:\\\\d\\{2\\}\\\\\\.000Z/,
    },
  },
}
`;
